### ISSUE TITLE
Sampler: actually save the samples

### ISSUE THREAD

**Issue by Bleuzen** [sadko4u/lsp-plugins#212]
Hi, so while using the LSP sampler plugin(s) I noticed something really bad. I work on two different machines and both have samples stored in different paths. One of them also does not have the complete sample collection.
Now if I create a project on machine A, I can't continue it on machine B, because the samples are in a different location.
This is a problem, because LSP Sampler Plugins only save the absolute path to the samples, but not the samples itself!
This is not only bad for one working on multiple machines, but also bad for archiving projects or sending them to a friend.
Please copy the samples over into the project.

For inspiration, here is how other plugins do it (assuming Ardour is used as DAW):

[samplv1](https://samplv1.sourceforge.io/):
When saving the project, it creates a symlink to the used sample in the projects externals directory:
![image](https://user-images.githubusercontent.com/12885163/142471504-6c296230-3528-4663-b810-abd94c750fb6.png)
While I would rather prefer a full copy, this is still acceptable, because Ardours archiving feature automatically resolves symlinks and creates copies of the used files for the archive.
This makes it a more future prove archive and allows sending the project to a friend or another machine, without them having to have the samples (and even in the exact same place).

[Geonkick](https://github.com/free-sm/geonkick):
This one saves the sample directly in the plugins `state.ttl` file as base64 encoded.
This way it is very easy to work on multiple machines because it includes everything needed. Just copy over the project directory and it "just works".
Only downside may be disc space usage. The problem is that everytime a change is made in plugin settings, the samples are saved again and since Ardour keeps a history of plugin states, this can create some duplicates.

**Comment by sadko4u** [comment]
> When saving the project, it creates a symlink to the used sample in the projects externals directory:

That's not correct. The symlink does the host. How does it handle the things with symlinks, it's the hotst's own implementation.
The plugins use LV2:State extension to load and store the settings. So when the LV2:restore event occurs, the plugins receive the path provided by the host to load samples.
I think @x42 can explain it more detailed.

**Comment by sadko4u** [comment]
Here what I've got after placing sampler to a new project, loading samples and saving the project:
```
sadko@sadko:/home/work/Ardour/test-sampler/plugins/267/state1> ls -l
lrwxrwxrwx 1 sadko users     34 ноя 19 02:29 Agogo-Hi-10.wav -> ../../../externals/Agogo-Hi-10.wav
lrwxrwxrwx 1 sadko users     35 ноя 19 02:29 Agogo-Low-03.wav -> ../../../externals/Agogo-Low-03.wav
-rw-r--r-- 1 sadko users    493 ноя 19 02:29 manifest.ttl
-rw-r--r-- 1 sadko users 108385 ноя 19 02:29 state.ttl
```

Now what about `externals`:
```
sadko@sadko:/home/work/Ardour/test-sampler/externals> ls -l
lrwxrwxrwx 1 sadko users 77 ноя 19 02:29 Agogo-Hi-10.wav -> ../../../../sadko/.hydrogen/data/drumkits/Forzee Mono Drumkit/Agogo-Hi-10.wav
lrwxrwxrwx 1 sadko users 78 ноя 19 02:29 Agogo-Low-03.wav -> ../../../../sadko/.hydrogen/data/drumkits/Forzee Mono Drumkit/Agogo-Low-03.wav
```

The plugin can not affect these paths since these paths are created by the host (in my case - Ardour 6.5.268).

**Comment by x42** [comment]
This is by design. Even if you use a sample multiple times in different plugins there is only one reference to that in the `externals/` folder. Which is both brilliant, yet inconvenient to users without unix experience.

It is really liblilv abstracting that (`lilv_state_save` calls `lilv_state_make_links`), the host does not have full control over this either. It's part of the plugin spec.

May I suggest to use `rsync` to copy the session? By default, it copies depending on the destination. If the link points to an absolute path outside the tree, the file is de-referenced. However if the link points to a relative path (e.g. `../../externals/`), it will continue pointing to that.

**Comment by Bleuzen** [comment]
> Here what I've got after placing sampler to a new project, loading samples and saving the project:

And here is mine, new project with new Midi track and Sampler loaded with a sample:

```
❯ pwd
/home/elias/Ardour/test
❯ ls plugins
❯ ls externals
~/Ardour/test ❯ 
```

In contrast to your result, both dirs (plugins and externals) are empty.
At this point I was wondering...
The only reference to the plugin I found was in the <project>.ardour file.
And since seeing this, I realized:
I thought I was using the LV2 version all the time, but Ardour loaded up the VST version.

Here is what I can choose when creating a new midi track:
![image](https://user-images.githubusercontent.com/12885163/142614881-abc5b4e5-0887-4766-8048-e495a271fbdc.png)

Usually, different formats of plugins are shown behind the name ("(VST)" or "(LV2)") but for LSP there is no format indication, so I thought it would be LV2.

So maybe the problem is only with the VST version of the Sampler. Will try further later.

**Comment by sadko4u** [comment]
Yes, VST may be the case since it has another mechanism of saving/restoring plugin state.
You can simply switch from VST to LV2 by placing one more plugin and doing export/import of the configuration via the UI.

**Comment by Bleuzen** [comment]
Hm Ardour doesn't give me the option to load the LV2 version of Sampler.
Here is what I get:
![image](https://user-images.githubusercontent.com/12885163/142616255-2bcff0e3-c8d2-481a-9fad-6af7204f60c0.png)

While the LSP LV2 plugins are installed and I also have the plugin as LV2 installed:
```
❯ ls -lah /usr/lib/lv2/lsp-plugins.lv2/sampler_stereo.ttl
-rw-r--r-- 1 root root 12K  1. Apr 2021  /usr/lib/lv2/lsp-plugins.lv2/sampler_stereo.ttl
```

Plugins were installed on machine A from arch repos:
https://archlinux.org/packages/community/x86_64/lsp-plugins/
and on machine B from flathub:
https://github.com/flathub/org.freedesktop.LinuxAudio.Plugins.LSP

But are the same, Ardour shows other LSP LV2s but no Sampler.
Even after rescanning plugins.
Not sure why.
Maybe a bug in Ardour? @x42

**Comment by sadko4u** [comment]
> Hm Ardour doesn't give me the option to load the LV2 version of Sampler.

There may be some reasons: bad build, missing dependencies for binaries, bug in plugins, bug in host.

> Usually, different formats of plugins are shown behind the name ("(VST)" or "(LV2)") but for LSP there is no format indication, so I thought it would be LV2.

The length of the VST2 plugin name is limited.
It is not always possible to add the `(VST)` postfix.

Oh, I've noticed you have native german translation.

Suchen Sie bitte das wort `Schlagzeug` in dem Dialog.

**Comment by Bleuzen** [comment]
Ah yeah, looks better:
![image](https://user-images.githubusercontent.com/12885163/142618553-aef022f6-a2b6-4b81-801d-728d36891c6e.png)

Only 12 and 24 instrument versions are shown but simple single sampler is missing. Isn't there a LV2 version?

**Comment by sadko4u** [comment]
For single sampler look for `Klangerzeuger`.
It's a feature of LV2 that it allows to provide plugin names with different translations and the host can take this translations depending on the system locale.
In 1.2.0 we're thinking on getting rid of german names of plugins in LV2 since it looks very annoying for native german speakers since... they are looking for english names of plugins.

**Comment by x42** [comment]
While you may find "Klangerzeugung" in some old publications.
"Klangerzeuger" is really unusual. It's not a word any German musician would use, except perhaps to make some joke.

**Comment by Bleuzen** [comment]
Oh well. The translation should really be removed. By that name, it could also be a Synth and it sounds wrong.

Anyway, gave the LV2 version a try.
The good news: It does create symlinks in the externals dir.

But compared to the VST version, the LV2 version is not registered as an instrument plugin.

1) I can not choose it when creating a new MIDI track in Ardour. This dropdown only contains the VST version:
![grafik](https://user-images.githubusercontent.com/12885163/142674317-d853c972-a978-4fcd-b505-5bc525b1a923.png)

2) While the VST version is clearly labeled as instrument:
![grafik](https://user-images.githubusercontent.com/12885163/142674631-4debc0a2-95a6-4d91-86d4-a25237069145.png)
The LV2 is not, it seems to be like an effect plugin:
![grafik](https://user-images.githubusercontent.com/12885163/142674791-6ecba0e6-8a99-46ea-ae86-87e0e5343cef.png)

3) While I can set the volume of the VST version with the fader, in the LV2 I can only set the midi velocity:
![grafik](https://user-images.githubusercontent.com/12885163/142674925-adaafa4b-c0fc-4f93-9dcc-da5557612a84.png)

**Comment by x42** [comment]
Re 3) Move the synth plugin pre-fader in Ardour's channel-strip.

**Comment by Bleuzen** [comment]
@x42 hm, doesn't change it:
![grafik](https://user-images.githubusercontent.com/12885163/142677895-1e84beb9-ada2-43e8-a12b-063da71eb5fe.png)

The track somehow has no audio output:
![grafik](https://user-images.githubusercontent.com/12885163/142678043-0c8fc589-9144-4ad4-826a-330cc8efb036.png)

But it does when loading another "real" instrument plugin.

I found that this can be worked around by disabling "Striktes E/A" on the track. This way, I can set two output channels and connect the plugin to output.

However this should happen automatically and also work with strict e/a.

**Comment by x42** [comment]
I see. The problem is that the plugin is not marked as Instrument.

When you use "strict I/O", Ardour does not allow a plugin to modify the channel-count. A mono track always remains mono, A MIDI track remains MIDI only, etc.

One excepts to that rule are MIDI instruments.

So until @sadko4u fixes the LV2 classification (to address your other two issues), you can work around this by either manually connecting the outputs or disabling "strict I/O".  The latter can be done when creating the track or later in the context-menu of the mixer-strip:

![image](https://user-images.githubusercontent.com/30246/142679097-dca34bf7-b176-418e-a0d4-cdd8cb2c7e19.png)

**Comment by sadko4u** [comment]
> I see. The problem is that the plugin is not marked as Instrument.

Hmmm. From the TTL file:

```
lsp:multisampler_x12
        a lv2:Plugin, doap:Project, lv2:InstrumentPlugin ;
```

@x42 what did I wrong?

**Comment by sadko4u** [comment]
It seems that it should be like this:
```
lsp:multisampler_x12
        a lv2:InstrumentPlugin, doap:Project ;
```

But... Why Ardour is still not handling multiple types of plugin instance?

**Comment by sadko4u** [comment]
Committed changes for LV2 TTL generator: 9b54330d1318d5616eb3393364cb01c83735cac2

**Comment by sadko4u** [comment]
Merged changes to the 'devel' branch: d475629b4b9317245c8264347a9980346caad219

**Comment by x42** [comment]
> But... Why Ardour is still not handling multiple types of plugin instance?

because there is only one category. http://drobilla.net/docs/lilv/#ga87106089cfc5122c4d5323cb20634ee8
compare to `lv2info`

**Comment by sadko4u** [comment]
> > But... Why Ardour is still not handling multiple types of plugin instance?
> 
> because there is only one category. http://drobilla.net/docs/lilv/#ga87106089cfc5122c4d5323cb20634ee8 compare to `lv2info`

@x42 how do you think, is it a good reason to make a feature request to lilv to support multiple plugin classes?

**Comment by sadko4u** [comment]
Available in 1.1.31 release!
